File: atcoder/python/abc210_219/abc210_D.py

import sys
infile = sys.stdin.buffer

def gs()  : return infile.readline().rstrip()
def gi()  : return int(gs())
def gf()  : return float(gs())
def gss() : return gs().split()
def gis() : return [int(x) for x in gss()]
def gfs() : return [float(x) for x in gss()]

def main(infn="") :
    global infile
    infile = open(infn,"r") if infn else open(sys.argv[1],"r") if len(sys.argv) > 1 else sys.stdin
    ## Run 2 DPS:  One looking back up and left, and One looking back down and left
    H,W,C = gis()
    A = []
    for i in range(H) : A.append(gis())
    inf = 10**18
    best = inf
    dp1 = [[inf]*W for _ in range(H)]
    dp2 = [[inf]*W for _ in range(H)]
    for i in range(H) :
        for j in range(W) :
            (opt1,opt2) = (inf,inf) if i == 0 else (A[i-1][j]+C,dp1[i-1][j]+C)
            (opt3,opt4) = (inf,inf) if j == 0 else (A[i][j-1]+C,dp1[i][j-1]+C)
            dp1[i][j] = min(opt1,opt2,opt3,opt4)
            cand = dp1[i][j] + A[i][j]
            best = min(best,cand)
    for i in range(H-1,-1,-1) :
        for j in range(W) :
            (opt1,opt2) = (inf,inf) if i == H-1 else (A[i+1][j]+C,dp1[i+1][j]+C)
            (opt3,opt4) = (inf,inf) if j == 0 else (A[i][j-1]+C,dp1[i][j-1]+C)
            dp1[i][j] = min(opt1,opt2,opt3,opt4)
            cand = dp1[i][j] + A[i][j]
            best = min(best,cand)
    sys.stdout.write(str(best)+'\n')

if __name__ == '__main__' :
    main()
    sys.stdout.flush()

File: abc/go/abc210/abc210_D/abc210_D.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)
var wrtr = bufio.NewWriterSize(os.Stdout, 10_000_000)
var rdr = bufio.NewScanner(os.Stdin)
func gs() string  { rdr.Scan(); return rdr.Text() }
func gi() int     { i,e := strconv.Atoi(gs()); if e != nil {panic(e)}; return i }
func gi3() (int,int,int) { return gi(),gi(),gi() }
func gis(n int) []int  { res := make([]int,n); for i:=0;i<n;i++ { res[i] = gi() }; return res }
func twodi(n int,m int,v int) [][]int {
	r := make([][]int,n); for i:=0;i<n;i++ { x := make([]int,m); for j:=0;j<m;j++ { x[j] = v }; r[i] = x }; return r
}
func min(a,b int) int { if a > b { return b }; return a }
func powint(a,e int) int { res, m := 1, a; for e > 0 { if e&1 != 0 { res = res * m }; m = m * m; e >>= 1 }; return res }
func main() {
	//f1, _ := os.Create("cpu.prof"); pprof.StartCPUProfile(f1); defer pprof.StopCPUProfile()
	defer wrtr.Flush()
	infn := ""; if infn == "" && len(os.Args) > 1 {	infn = os.Args[1] }
	if infn != "" {	f, e := os.Open(infn); if e != nil { panic(e) }; rdr = bufio.NewScanner(f) }
	rdr.Split(bufio.ScanWords); rdr.Buffer(make([]byte,1024),1_000_000_000)
	// PROGRAM STARTS HERE
	H,W,C := gi3(); A := make([][]int,H)
	for i:=0;i<H;i++ { A[i] = gis(W) }
	inf := powint(10,18); dp1,dp2 := twodi(H,W,inf),twodi(H,W,inf)
	// Pass 1, looking up and to the left
	ans := inf
	for i:=0;i<H;i++ {
		for j:=0;j<W;j++ {
			best := inf
			if i > 0 { best = min(best,C + min(dp1[i-1][j],A[i-1][j])) }
			if j > 0 { best = min(best,C + min(dp1[i][j-1],A[i][j-1])) }
			dp1[i][j] = best
			ans = min(ans,A[i][j]+best)
		}
		for j:=W-1;j>=0;j-- {
			best := inf
			if i > 0   { best = min(best,C + min(dp2[i-1][j],A[i-1][j])) }
			if j < W-1 { best = min(best,C + min(dp2[i][j+1],A[i][j+1])) }
			dp2[i][j] = best
			ans = min(ans,A[i][j]+best)
		}
	}
	fmt.Println(ans)
}

